### PR TITLE
Bring to CoreCLR the SecureString implementation from CoreFx.

### PR DESCRIPTION
**Pull request by AlexGhiondea**
Once this is checked in there is work remaining in CoreFx:
- Remove the SecureString implementation from CoreFx and type-forward to this
- Remove the implementation of CryptographicException and type-forward to this

### CODE REVIEW

File: src/mscorlib/corefx/System/Security/SecureString.Windows.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            AllocateBuffer(str._buffer.Length);
+            SafeBSTRHandle.Copy(str._buffer, _buffer, str._buffer.Length * sizeof(char));
+
+            _decryptedLength = str._decryptedLength;
+            _encrypted = str._encrypted;
+        }
+
+        private unsafe void InitializeSecureString(char* value, int length)
+        {
+            BCLDebug.Assert(length >= 0, $"Expected non-negative length, got {length}");
```
Comment (JeremyKuhne) [inline src/mscorlib/corefx/System/Security/SecureString.Windows.cs]: Just use Debug.Assert

Comment (AlexGhiondea) [inline src/mscorlib/corefx/System/Security/SecureString.Windows.cs]: Ah - yes! When I originally ported this, your change was not there :).

Will change it!

File: src/mscorlib/corefx/System/Security/SecureString.Unix.cs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private void EnsureCapacity(int capacity)
+        {
+            // Make sure the requested capacity doesn't exceed SecureString's defined limit
+            if (capacity > MaxLength)
+            {
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
```
Comment (justinvp) [inline src/mscorlib/corefx/System/Security/SecureString.Unix.cs]: Nit: `nameof(capacity)`? (I see there are some other uses of `nameof` in this change, so presumably this language feature can be used from corelib.)

File: src/mscorlib/model.xml [hunk 3]
```diff
@@ -7192,6 +7192,21 @@
...
+      <Member Name="AppendChar(System.Char)" />
+      <Member Name="Clear" />
+      <Member Name="Copy" />
+      <Member Name="Dispose" />
+      <Member Name="InsertAt(System.Int32,System.Char)" />
+      <Member Name="IsReadOnly" />
+      <Member Name="MakeReadOnly" />
+      <Member Name="RemoveAt(System.Int32)" />
+      <Member Name="SetAt(System.Int32,System.Char)" />
+    </Type>
```
Comment (stephentoub) [inline src/mscorlib/model.xml]: What about SecureStringMarshal?

File: src/mscorlib/corefx/System/Security/CryptographicException.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+using System.Runtime.Serialization;
+
+namespace System.Security.Cryptography
+{
+    [Serializable]
+    public class CryptographicException : SystemException
```
Comment (stephentoub) [inline src/mscorlib/corefx/System/Security/CryptographicException.cs]: So are we now exposing CryptographicException from coreclr as well?  Presumably it needs to be added to model.xml as well then?

Comment (AlexGhiondea) [inline src/mscorlib/corefx/System/Security/CryptographicException.cs]: I added them both to model.xml.

**Pull request merged**